src/rgw/rgw_lua.cc

```diff
@@ -175,42 +190,106 @@ int install_packages(const DoutPrefixProvider *dpp, rgw::sal::Driver* driver,
     return 0;
   }
   if (ret < 0) {
-    output.append("failed to get lua package list");
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to get package list. error: " << ret << dendl;
     return ret;
   }
   // verify that luarocks exists
   const auto p = bp::search_path("luarocks");
   if (p.empty()) {
-    output.append("failed to find luarocks");
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to find luarocks" << dendl;
     return -ECHILD;
   }
 
+  // create a temporary directory to install all luarocks packages
+  std::string tmp_path_template = fs::temp_directory_path();
+  tmp_path_template.append("/luaXXXXXX");
+  const auto tmp_luarocks_path = mkdtemp(tmp_path_template.data());
+  if (!tmp_luarocks_path) { 
+    const auto rc = -errno;
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to create temporary directory from template: " << 
+      tmp_path_template << ". error: " << rc << dendl;
+    return rc;
+  }
+
   // the lua rocks install dir will be created by luarocks the first time it is called
   for (const auto& package : packages) {
     bp::ipstream is;
-    const auto cmd = p.string() + " install --lua-version " + CEPH_LUA_VERSION + " --tree " + luarocks_path + " --deps-mode one " + package;
+    const auto cmd = p.string() + " install --lua-version " + CEPH_LUA_VERSION + " --tree " + tmp_luarocks_path + " --deps-mode one " + package;
     bp::child c(cmd, bp::std_in.close(), (bp::std_err & bp::std_out) > is);
 
     // once package reload is supported, code should yield when reading output
-    std::string line = std::string("CMD: ") + cmd;
+    std::string lines = std::string("Lua CMD: ") + cmd;
+    std::string line;
 
     do {
       if (!line.empty()) {
-        output.append(line);
-        output.append("\n");
+        lines.append("\n\t");
+        lines.append(line);
       }
     } while (c.running() && std::getline(is, line));
 
     c.wait();
     if (c.exit_code()) {
       failed_packages.insert(package);
     }
+    ldpp_dout(dpp, 20) << lines << dendl;
+  }
+  
+  // luarocks directory cleanup
+  std::error_code ec;
+  if (fs::remove_all(luarocks_path, ec)
+      == static_cast<std::uintmax_t>(-1) &&
+      ec != std::errc::no_such_file_or_directory) {
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to clear luarocks directory: " <<
+      luarocks_path << ". error: " << ec.message() << dendl; 
+    return -ec.value();
+  }
+  auto rc = create_directory_p(dpp, luarocks_path);
+  if (rc < 0) {
+    ldpp_dout(dpp, 1) << "Lua ERROR: failed to recreate luarocks directory: " <<
+      luarocks_path << ". error: " << rc << dendl; 
+    return rc;
+  }
+
+  // switch temporary install directory to luarocks one
+  fs::rename(tmp_luarocks_path, luarocks_path, ec);
```

Comment:
> the rename will only work if both paths are in the same filesystem. is there any guarantee that `luarocks_path` will also be under /tmp?

currently, we don't have that guarantee, but maybe we can add that? we install packages on restart, so a machine restart will cause reintsall anyway

> 
> maybe it would be better to just rerun `install_packages()` in place without trying to do any clean up?

this would requre more complex and error prone implementations, since we would need to do a diff and uninstall packages that are not in the allowlist

Reply:
> this would requre more complex and error prone implementations, since we would need to do a diff and uninstall packages that are not in the allowlist

is that really required? if it's in a tmp directory they'll go away eventually. and if we require each radosgw to maintain its own separate directory, it could do that that cleanup on shutdown to handle lua paths outside of /tmp

Reply:
> > this would requre more complex and error prone implementations, since we would need to do a diff and uninstall packages that are not in the allowlist
> 
> is that really required? if it's in a tmp directory they'll go away eventually. and if we require each radosgw to maintain its own separate directory, it could do that that cleanup on shutdown to handle lua paths outside of /tmp

do you mean do a delete+copy instead of move?

Reply:
> do you mean do a delete+copy instead of move?

i mean running install_packages() on the existing path without deleting anything first. is that something luarocks would handle correctly if package versions were upgraded?

otherwise, to make the `rename()` work, you could use subdirectories under `luarocks_path` to make sure they end up on the same filesystem. but the rmdir->mkdir->rename approach is not atomic. you might just keep installing to a new subdirectory each time, then switch what rgw sees as the `luarocks_path` to point at the newest one

Reply:
> i mean running install_packages() on the existing path without deleting anything first. is that something luarocks would handle correctly if package versions were upgraded?

I'll have to test that, but might be trickey with packages that has conflicts and dependencies - this is why I wanted to avoid that

> otherwise, to make the rename() work, you could use subdirectories under luarocks_path to make sure they end up on the same filesystem

looks to me like a good option

> but the rmdir->mkdir->rename approach is not atomic

is the concern that 2 reloads will be done at the same time? or that one of the steps would fail?